Optimize pillow memory usage: do not load cache data sources for inactive domains

millerdev opened the pull request:
This is a major refactoring of `UcrTableManager` that optimizes caching behavior so inactive data source configurations are not loaded into memory. UCR data source configurations are loaded as cases are processed for the relevant domain. Each UCR table adapter, which contains the associated data source configuration, is cached in memory for 2 days after the most recent pillow activity that references it, after which point it will be purged from the cache.

https://dimagi.atlassian.net/browse/SAAS-18100

:blowfish: Review by commit.

## Safety Assurance

### Safety story

New tests were added for all new components. Existing tests were updated to accommodate the new table manager structure. The code will be run on staging for some time before deploying to production.

### Automated test coverage

Yes.

### QA Plan

https://dimagi.atlassian.net/browse/SAAS-18321

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

millerdev committed:
Move get_all_configs into get_filtered_configs

millerdev committed:
Eliminate UcrTableManager.relevant_domains

millerdev committed:
Add argument to fetch data sources by domain

millerdev committed:
Add config_id and is_active to IndicatorAdapter

millerdev committed:
Add TTLCache and MigrationCache for UCR pillows

These caches will allow pillows to run with better memory efficiency.
For non-migration processes, they keep only recently used (rather than
all) UCR data source adapters in memory. The migration process still
loads all data source adapters into memory.

Currently these are unused. A subsequent commit will integrate them into
corehq/apps/userreports/pillow.py

millerdev committed:
Add TaskCoordinator for UCR table rebuilds

Currently this is unused. A subsequent commit will integrate it into
corehq/apps/userreports/pillow.py

millerdev committed:
Fix tests and move test code out of production code

millerdev committed:
Refactor: remove unused parameter

millerdev committed:
Lint

millerdev committed:
Do not exclude migrating domains from UCR rebuilds

This was originally intended to exclude domains that were migrating to a
different HQ environment[1], but since then DomainMigrationProgress has
been adopted for many other kinds of migrations, most of which are not
relevant to UCR rebuilds.

No DomainMigrationProgress records exist on prod pertaining to migrating
to a different environment, and set_migration_started(...) is not used
to start that type of migration anywhere in code, so it is unclear how
this model was used for that purpose historically. A better
implementation would have checked for migrations with a specific slug,
but removal seems justified since that slug is unknown.

[1] https://github.com/dimagi/commcare-hq/pull/28532

millerdev committed:
Use TaskCoordinator + TTLCache in UcrTableManager

Instead of keeping all UCR table adapters in memory constantly, this
will load adapters only when a change is received for a relevant domain.
Adapters will be cached in memory for 48 hours after a change is
received for the relevant domain.

millerdev committed:
Comments on (stale) data source config caching

millerdev committed:
Use MigrationCache for dedicated migration process

millerdev commented on corehq/apps/userreports/pillow.py:
```diff
@@ -243,73 +307,16 @@ def get_filtered_configs(self, configs=None):
-            for source in provider.get_data_sources_modified_since(timestamp)
-        ]
-        filtered_data_sources = self.get_filtered_configs(new_data_sources)
-        invalid_data_sources = {ds._id for ds in new_data_sources} - {ds._id for ds in filtered_data_sources}
```
This never worked because the values in `invalid_data_sources` were expected to be _data source config_ objects, not ids.

millerdev commented on corehq/apps/userreports/tests/test_registry_pillow.py:
```diff
@@ -35,95 +34,82 @@ def setUpTestData(cls):
+    def test_bootstrap(self):
         self._bootstrap_manager_with_data_source()
 
-    def test_ignore_migrating_domains(self, mock_domains_with_migrations):
```
See https://github.com/dimagi/commcare-hq/pull/36819/commits/8edc65b295a0c783522d5118aac75497739ebe15

millerdev committed:
Merge remote-tracking branch 'origin/master' into dm/ucr-pillow-ttl-cache

kaapstorm commented on corehq/apps/userreports/pillow.py:
```diff
@@ -125,79 +128,133 @@ class UcrTableManager(ABC):
+        rebuild_sql_tables(adapters)
+
+    def refresh_cache(self):
+        self.cache.refresh()
```
I see that `ConfigurableReportPillowProcessor` calls `refresh_cache()` in its `process_change()` and `process_changes_chunk()` methods. How often should `refresh_cache()` ideally be called? Would it be over-engineering to track requests to refresh the cache, and refresh it _up to_ once a minute? (i.e. If `refresh_cache()` was called less than a minute ago, ignore the request.) ... I also don't know if that's safe.

millerdev replied:
Safety is the main concern. As is, it's not changing behavior from how it as before.

gherceg reviewed (commented):
Looks good. Planning to take another pass on https://github.com/dimagi/commcare-hq/pull/36819/commits/27f2552eff2a3ffeadb3100920cb64765021dff1.

millerdev committed:
Use generators rather than allocating a new list for each filter

Unclear if performance or memory footprint is improved (didn't measure),
but generators are so concise. They make the code easier to read.

millerdev commented on corehq/apps/userreports/pillow.py:
```diff
@@ -97,10 +94,9 @@ def _filter_domains_to_skip(configs):
-    migrating_domains = all_domains_with_migrations_in_progress()
     return [
         config for config in configs
-        if config.domain not in migrating_domains and (not config.is_static or config.domain in existing_domains)
```
What migration slug do we use for that kind of migration?

millerdev committed:
Remove old comment. The code is clear without it.

millerdev commented on corehq/apps/userreports/pillow.py:
```diff
@@ -544,8 +534,6 @@ def process_change(self, change):
 
         with TimingContext() as timer:
             eval_context = EvaluationContext(doc)
-            # make copy to avoid modifying list during iteration
```
I think this comment was outdated. I don't see where the list is modified.

millerdev commented on corehq/apps/userreports/tests/test_adaptercache.py:
```diff
@@ -0,0 +1,220 @@
+        with freeze_time('2020-01-04T01:00:00Z'):
+            cache.refresh()
+            assert 'two' not in cache.adapters
+            assert cache.get_adapters('one') != result
```
Correct.

millerdev committed:
Improve test clarity

millerdev commented on corehq/apps/userreports/tests/utils.py:
```diff
@@ -37,6 +39,23 @@ def get_sample_registry_data_source(**kwargs):
 
 
+def bootstrap_pillow(pillow, *configs, rebuild_adapters=False):
+    configs_by_domain = {}
+    for config in configs:
+        configs_by_domain.setdefault(config.domain, []).append(config)
```
In this case it was intentional. I didn't want the automatic value initialization behavior of `defaultdict` because it could be surprising in far-away places like `MockDataSourceProvider`. Maybe that's overly conservative on my part?

millerdev commented on corehq/apps/userreports/pillow.py:
```diff
@@ -212,14 +235,19 @@ def __init__(self, data_source_providers, ucr_division=None,
+            for config in self.get_filtered_configs(configs):
+                yield (config.domain, config)
 
+    def get_filtered_configs(self, configs):
```
f47f94c55c13bb4840de669c310e3323a607493e

millerdev commented on corehq/apps/userreports/tests/test_adaptercache.py:
```diff
@@ -0,0 +1,220 @@
+            cache.refresh()
+            assert cache.get_adapters('one') == result  # no change
+            assert cache.get_adapters('two') == [adapter1, adapters['two']]
+            assert 'three' not in cache.adapters
```
61ec02a07e33c6885c7193084535e2d63fbd1734

millerdev commented on corehq/apps/userreports/tests/test_adaptercache.py:
```diff
@@ -0,0 +1,220 @@
+            assert cache.get_adapters('two') == result
+
+        with freeze_time('2020-01-02T00:00:00Z'):
+            adapter1.modified = datetime.now(UTC)
```
Attempted to improve clarity in 61ec02a07e33c6885c7193084535e2d63fbd1734

millerdev commented on corehq/apps/userreports/tests/test_adaptercache.py:
```diff
@@ -0,0 +1,220 @@
+            adapter = Adapter('one')
+            cache = TTLCache(iter_adapters)
+            result = cache.get_adapters('one')
+            assert result == [adapter]
```
61ec02a07e33c6885c7193084535e2d63fbd1734

millerdev committed:
Merge branch 'master' into dm/ucr-pillow-ttl-cache

millerdev committed:
Replace freezegun with time-machine
